qgis/QGIS_PluginProperWay.md

```diff
@@ -0,0 +1,230 @@
+- **Defensive coding**: Meaning, always think what could go wrong when executing your code and properly guard against user-induced errors. This is very important for QGIS plugin development as users are often technically rather casual
+- **Unit and e2e testing**: Unit testing makes sure single methods are doing what they're supposed to, while e2e (or integration) tests rather test an entire workflow start to finish. In our case an e2e test would be to test whether a layer was created after the user hit OK with valid coordinates.
+
+As most QGIS plugins are openly accessible, a smart plugin maintainer would like community contributions for bug fixes or new features. Following best practices like the ones above makes community contribution much more likely. No one wants to dive into spaghetti code.
+
+## 5 - Project structure
+
+In terms of QGIS plugins, a sensible project structure usually emulates QGIS' own package structure, where you'd end up with something like this:
+
+```
+quick_api_interactive_proper
+├── .git/
+├── dist/
+├── quick_api/
+│   ├── core/
+│   ├── gui/
+│   ├── icons/
+│   ├── ui/
+│   ├── __init__.py
+│   ├── metadata.txt
+│   └── quick_api.py
+├── tests/
+├── .gitignore
+├── .pre-commit-config.yaml
+├── README.md
+└── setup.cfg
+```
+
+First, the actual plugin code should live as a package in your Git folder and it's common to give it the same name as the plugin, i.e. `quick_api/`. That way you'll only package relevant files for the plugin and can omit project/development files like README and dev setup files (e.g. `setup.cfg`). When you package your plugin for QGIS (i.e. zip it up), you can put the ZIP(s) in the `dist/` folder.
+
+The code inside the plugin package is commonly leaning towards QGIS own logic:
+- `core/`: Contains the business logic of your plugin, mostly functionality that doesn't depend on the rest of your plugin like the UI. In our case that'd would the logic to request Nominatim and the map tool capturing clicked points.
+- `gui`: Contains the logic around your dialogs, which will make use of the `core/` functionality.
+- `icons`: A place to hold your plugin-specific icons, doesn't need to be a Python package (i.e. no need for a `__init__.py`).
+- `ui`: Contains the raw and/or compiled UI files.
+- `processing` (if needed): Contains the modules used to integrate with QGIS Processing framework.
+- `__init__.py`: **Only** contains the `classFactory()` QGIS expects every plugin to have.
+- `quick_api.py`: The class QGIS uses to instantiate the plugin. This should only contain functionality which is important to the **main** QGIS app, such as adding toolbars/menus (e.g. in `initGui()`) or the `unload()` function. See our [QGIS Plugin 101 tutorial](https://gis-ops.com/qgis-3-plugin-tutorial-plugin-development-reference-guide/) for more info.
+
+## 6 - Linters
```

Comment:
Would also be nice to add a code formatter which runs before the linter, i.e. black, I think it works quite well together with flake8, https://ljvmiranda921.github.io/notebook/2018/06/21/precommits-using-black-and-flake8/

Reply:
yeah, I thought about using yapf (though not sure about compat with flake8)… Maybe it's time to give black a chance..

Reply:
added